sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/GpuHiveSimpleUDF.scala

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.hive.rapids
+
+import ai.rapids.cudf.{NvtxColor, NvtxRange}
+import com.nvidia.spark.RapidsUDF
+import com.nvidia.spark.rapids.{ExprChecks, ExprMeta, ExprRule, GpuColumnVector, GpuExpression, GpuOverrides, GpuScalar, RepeatingParamCheck, TypeSig}
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
+import org.apache.hadoop.hive.ql.exec.UDF
+
+import org.apache.spark.sql.catalyst.expressions.{Expression, UserDefinedExpression}
+import org.apache.spark.sql.hive.HiveShim.HiveFunctionWrapper
+import org.apache.spark.sql.hive.HiveSimpleUDF
+import org.apache.spark.sql.types.DataType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+/** GPU-accelerated version of Spark's `HiveSimpleUDF` */
+case class GpuHiveSimpleUDF(
+    name: String,
+    funcWrapper: HiveFunctionWrapper,
+    children: Seq[Expression],
+    dataType: DataType,
+    udfDeterministic: Boolean) extends GpuExpression with UserDefinedExpression {
+
+  private[this] val nvtxRangeName = "UDF: " + name
+
+  override lazy val deterministic: Boolean = udfDeterministic && children.forall(_.deterministic)
+
+  override def nullable: Boolean = true
+
+  override def foldable: Boolean = udfDeterministic && children.forall(_.foldable)
+
+  @transient
+  lazy val function: RapidsUDF = funcWrapper.createFunction[UDF]().asInstanceOf[RapidsUDF]
+
+  override def toString: String = {
+    s"$nodeName#${funcWrapper.functionClassName}(${children.mkString(",")})"
+  }
+
+  override def prettyName: String = name
+
+  override def sql: String = s"$name(${children.map(_.sql).mkString(", ")})"
+
+  private[this] def evalExpr(expr: Expression, batch: ColumnarBatch): GpuColumnVector = {
+    expr.columnarEval(batch) match {
+      case v: GpuColumnVector => v
+      case other =>
+        withResource(GpuScalar.from(other, expr.dataType)) { s =>
+          GpuColumnVector.from(s, batch.numRows(), expr.dataType)
+        }
+    }
+  }
+
+  override def columnarEval(batch: ColumnarBatch): Any = {
+    withResource(children.safeMap(evalExpr(_, batch))) { exprResults =>
+      val funcInputs = exprResults.map(_.getBase()).toArray
+      withResource(new NvtxRange(nvtxRangeName, NvtxColor.PURPLE)) { _ =>
+        GpuColumnVector.from(function.evaluateColumnar(funcInputs: _*), dataType)
```

Comment:
Shall we need to check the length and type of the cuDF ColumnVector returned from `evaluateColumnar` ?
Suppose it should have the same length with the input columns, and can cast to the `dataType` .

Reply:
Yes, that's a good point.  I added both row checks and type conversion checks.  There was already a conversion check occurring in `GpuColumnVector.from` but it only has an effect if assertions are enabled.  Since this is calling user-provided code, I added a version that performs an unconditional check.  cc: @revans2 for his thoughts on this.

Reply:
Table also will verify that the lengths of all of the columns are the same, but that might not happen until after the processing has completed. I agree checking these sooner than later is ideal, especially because it is user code that we will be dealing with.